### PR TITLE
Quick fix to the walk method where an infinite loop can occur

### PR DESCRIPTION
**Pull request by dnesbitt61**
(on a single edge that is disconnected from everything). Change log warnings to debug and
add some simple stats to see what the success rate is. Add a lock around writing tiles. Update check_access to use a simple allow_edge_pred method that ultimately should match that used within OSMLR creation.

### CODE REVIEW

File: src/valhalla_associate_segments.cc [hunk 1]
```diff
@@ -482,7 +469,18 @@ std::vector<vb::GraphId> walk(const vb::PathLocation &origin, const vb::PathLoca
       //get the next edge
       edges.push_back(next_edge(edges.back(), reader, tile));
       //if the edge is invalid we have no where to go
-    } while(edges.back().Is_Valid());
+    } while(edges.back().Is_Valid() && edges.size() < 1000);
```
Comment (kevinkreiser) [inline src/valhalla_associate_segments.cc]: could we be a bit less blunt here? like use the length instead:

```diff
diff --git a/src/valhalla_associate_segments.cc b/src/valhalla_associate_segments.cc
index 796dabe3..1fdad03f 100644
--- a/src/valhalla_associate_segments.cc
+++ b/src/valhalla_associate_segments.cc
@@ -369,13 +369,14 @@ edge_association::edge_association(const bpt::ptree &pt)
   , m_tile(nullptr) {
 }
 
-vb::GraphId next_edge(const GraphId& edge_id, vb::GraphReader& reader, const vb::GraphTile*& tile) {
+vb::GraphId next_edge(const GraphId& edge_id, vb::GraphReader& reader, const vb::GraphTile*& tile, uint32_t& current_length) {
   //walk from this edge to the next one if there is only one choice of where
   //to walk if there are more choices then just return invalid to signify
   //stopping this is trying to mimic what osmlr generation does
   if(tile->id() != edge_id.Tile_Base())
     tile = reader.GetGraphTile(edge_id);
   const auto* edge = tile->directededge(edge_id);
+  current_length += edge->length();
   if(tile->id() != edge->endnode().Tile_Base())
     tile = reader.GetGraphTile(edge->endnode());
   const auto* node = tile->node(edge->endnode());
@@ -461,7 +462,7 @@ vb::GraphId edge_association::find_common_edge(
 }
 
 std::vector<vb::GraphId> walk(const vb::PathLocation &origin, const vb::PathLocation &dest,
-                              vb::GraphReader& reader, const vb::GraphTile* tile) {
+                              vb::GraphReader& reader, const vb::GraphTile* tile, uint32_t total_length) {
   //check for the easy case
   for (const auto &origin_edge : origin.edges)
     for (const auto &dest_edge : dest.edges)
@@ -471,6 +472,7 @@ std::vector<vb::GraphId> walk(const vb::PathLocation &origin, const vb::PathLoca
   //see if we can easily find a longer path
   for (const auto &origin_edge : origin.edges) {
     //try walking from here
+    uint32_t current_length = 0;
     std::vector<vb::GraphId> edges{origin_edge.id};
     do {
       //for each ending edge
@@ -480,9 +482,9 @@ std::vector<vb::GraphId> walk(const vb::PathLocation &origin, const vb::PathLoca
           return edges;
       }
       //get the next edge
-      edges.push_back(next_edge(edges.back(), reader, tile));
+      edges.push_back(next_edge(edges.back(), reader, tile, current_length));
       //if the edge is invalid we have no where to go
-    } while(edges.back().Is_Valid());
+    } while(edges.back().Is_Valid() && current_length < total_length * 1.1f);
   }
 
   //fail try a route?
@@ -584,7 +586,7 @@ std::vector<vb::GraphId> edge_association::match_edges(const pbf::Segment &segme
   std::remove_if(dest.edges.begin(), dest.edges.end(), [](const vb::PathLocation::PathEdge& e){return e.begin_node();});
 
   // check if its a trivial path between edges
-  auto walked_edges = walk(origin, dest, m_reader, m_tile);
+  auto walked_edges = walk(origin, dest, m_reader, m_tile, total_length);
   if (walked_edges.size()) {
     // TODO: check bearing, length, FRC, FOW, etc...
```

File: src/valhalla_associate_segments.cc [hunk 2]
```diff
@@ -844,14 +842,22 @@ void edge_association::add_tile(const std::string &file_name) {
...
+      } else {
+        failure_count++;
+      }
     }
     entry_id += 1;
   }
 
   //finish this tile
+  lock.lock();
   m_reader.Clear();
```
Comment (kevinkreiser) [inline src/valhalla_associate_segments.cc]: the graph reader is local so the clear doesnt need to be inside of the lock

**Comment by kevinkreiser** [comment]
looks good other than comments

**Comment by kevinkreiser** [comment]
this looks good to me thanks for the fix. quick question, the locking wasnt needed because we werent changing the actual contents (edges and nodes) of the tiles just the traffic association, and it wouldnt matter if it were there or not when read from another thread?

**Comment by dnesbitt61** [comment]
Yes, that is why I think locking is not needed. No issues seen with 4 threads on a larger data set.

**Pull request merged**